test/system/src/test/java/io/pravega/test/system/MultiReaderWriterWithFailOverTest.java

```diff
@@ -0,0 +1,427 @@
+import io.pravega.client.admin.impl.StreamManagerImpl;
+import io.pravega.client.stream.EventStreamReader;
+import io.pravega.client.stream.EventStreamWriter;
+import io.pravega.client.stream.EventWriterConfig;
+import io.pravega.client.stream.ReaderConfig;
+import io.pravega.client.stream.ReaderGroupConfig;
+import io.pravega.client.stream.ReinitializationRequiredException;
+import io.pravega.client.stream.ScalingPolicy;
+import io.pravega.client.stream.StreamConfiguration;
+import io.pravega.client.stream.impl.ClientFactoryImpl;
+import io.pravega.client.stream.impl.Controller;
+import io.pravega.client.stream.impl.ControllerImpl;
+import io.pravega.client.stream.impl.JavaSerializer;
+import io.pravega.common.concurrent.FutureHelpers;
+import io.pravega.test.system.framework.Environment;
+import io.pravega.test.system.framework.SystemTestRunner;
+import io.pravega.test.system.framework.services.BookkeeperService;
+import io.pravega.test.system.framework.services.PravegaControllerService;
+import io.pravega.test.system.framework.services.PravegaSegmentStoreService;
+import io.pravega.test.system.framework.services.Service;
+import io.pravega.test.system.framework.services.ZookeeperService;
+import mesosphere.marathon.client.utils.MarathonException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import java.util.TreeSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import lombok.extern.slf4j.Slf4j;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.stream.Collectors;
+
+@Slf4j
+@RunWith(SystemTestRunner.class)
+public class MultiReaderWriterWithFailOverTest {
+    private static final String STREAM_NAME = "testMultiReaderWriterStream";
+    private static final int NUM_WRITERS = 20;
+    private static final int NUM_READERS = 20;
+    private AtomicBoolean stopReadFlag;
+    private AtomicBoolean stopWriteFlag;
+    private AtomicLong eventData;
+    private AtomicLong eventReadCount;
+    private Long eventWriteCount1;
+    private Long eventWriteCount2;
+    private Long eventWriteCount3;
+    private Long eventWriteCount4;
+    private Long eventReadCount1;
+    private Long eventReadCount2;
+    private Long eventReadCount3;
+    private Long eventReadCount4;
+    private ConcurrentLinkedQueue<Long> eventsReadFromPravega;
+    private Service controllerInstance = null;
+    private Service segmentStoreInstance = null;
+    private URI controllerURIDirect = null;
+    private final Object waitCondition = new Object();
```

Comment:
Replace waitcondition with a reusable Latch, a semephore, or something similar. That would avoid the need to synchronize and would prevent the race condition where the notification comes before the wait.